core/lib/dal/src/protocol_versions_dal.rs

```diff
@@ -313,20 +322,86 @@ impl ProtocolVersionsDal<'_, '_> {
         })
     }
 
-    pub async fn last_version_id(&mut self) -> DalResult<Option<ProtocolVersionId>> {
-        Ok(sqlx::query!(
+    pub async fn get_patch_versions_for_vk(
+        &mut self,
+        minor_version: ProtocolVersionId,
+        recursion_scheduler_level_vk_hash: H256,
+    ) -> DalResult<Vec<VersionPatch>> {
+        let rows = sqlx::query!(
             r#"
             SELECT
-                MAX(id) AS "max?"
+                patch
             FROM
-                protocol_versions
+                protocol_patches
+            WHERE
+                minor = $1
+                AND recursion_scheduler_level_vk_hash = $2
+            ORDER BY
+                patch DESC
+            "#,
+            minor_version as i32,
+            recursion_scheduler_level_vk_hash.as_bytes()
+        )
+        .instrument("get_patch_versions_for_vk")
+        .fetch_all(self.storage)
+        .await?;
+        Ok(rows
+            .into_iter()
+            .map(|row| VersionPatch(row.patch as u32))
+            .collect())
+    }
+
+    /// Returns first patch number for the minor version.
+    /// Note, that some patch numbers can be skipped, so the result is not always 0.
+    pub async fn first_patch_for_version(
+        &mut self,
+        version_id: ProtocolVersionId,
+    ) -> DalResult<Option<VersionPatch>> {
+        let row = sqlx::query!(
+            r#"
+            SELECT
+                patch
+            FROM
+                protocol_patches
+            WHERE
+                minor = $1
+            ORDER BY
+                patch
+            LIMIT
+                1
+            "#,
+            version_id as i32,
+        )
+        .instrument("first_patch_for_version")
+        .fetch_optional(self.storage)
+        .await?;
+        Ok(row.map(|row| VersionPatch(row.patch as u32)))
+    }
+
+    pub async fn latest_semantic_version(&mut self) -> DalResult<Option<ProtocolSemanticVersion>> {
```

Comment:
when people deploy this binary version on their existing chain - what will be present in protocol_patches? would that table be empty? (if yes - then this will return an error / None, right?)

Reply:
table will have same number of rows as `protocol_versions`, there will patch 0 for each minor version